File: 22. CountSubsetsWithSum.py
# Topic: Count subsets with given sum using bitmasking

def count_subsets_with_sum(arr, target):
    n = len(arr); cnt = 0
    for mask in range(1<<n):
        s = sum(arr[i] for i in range(n) if mask & (1<<i))
        if s == target: cnt += 1
    return cnt

if __name__ == '__main__':
    print('Count sum=3:', count_subsets_with_sum([1,2,3],3))

# Time Complexity: O(2^n * n)
# Space Complexity: O(1)

File: 22. CountSubsetsWithSum.cpp
// Topic: Count subsets with given sum using bitmasking
#include <bits/stdc++.h>
using namespace std;

int countSubsetsWithSum(const vector<int>& arr, int target)
{
    int n = arr.size(), cnt = 0;
    for (int mask = 0; mask < (1 << n); ++mask)
    {
        int s = 0;
        for (int i = 0; i < n; ++i)
            if (mask & (1 << i))
                s += arr[i];
        if (s == target)
            cnt++;
    }
    return cnt;
}

int main()
{
    vector<int> a = {1, 2, 3};
    cout << "Count sum=3: " << countSubsetsWithSum(a, 3) << '\n';
    return 0;
}

// Time Complexity: O(2^n * n)
// Space Complexity: O(1)
